src/librustc/traits/project.rs

```diff
@@ -818,55 +822,74 @@ fn project_type<'cx, 'gcx, 'tcx>(
                                        &obligation_trait_ref,
                                        &mut candidates);
 
+    let decide_commit = |candidates: &mut ProjectionTyCandidateSet<'tcx>| {
```

Comment:
I have no way but to reorder this before the error handling. It looks non intuitive.